### ISSUE TITLE
Multipass problem Support

### ISSUE THREAD

**Issue by mzuenni** [DOMjudge/domjudge#2307]
### Description of the enhancement request
In multipass problems the submission gets started multiple times and is fed with the output of the output valdiator from the previous run (but besides that the new run has no knowledge of any prior run). This allows some new types of problems about information theory and some other fields.

### Expected behaviour
- The submission is run on the test input (like for normal problems)
- The output Validator is run on the submission output (like for normal problems)
- The output Validator either accepts, rejects or indicates that the submission should be run again with a new input that it provides

### Any other information that you want to share?
Multipass seems to also get added to the kattis problem package format: https://www.kattis.com/problem-package-format/spec/2023-07-draft.html#multi-pass-validation
Maybe it would also be nice to run the output validator with an argument which indicates which iteration this is or provide a way for the output validator to keep information between invocations.

**Comment by vmcj** [comment]
@mzuenni as discussed in person, I'll close this issue as this requires the problemspec to first have a definition for this. Please re-open when that is in a reasonable state so we can start implementing this.

**Comment by mzuenni** [comment]
Just a small update:

> A multi-pass validator can be used for problems that should run the submission multiple times sequentially, using a new input generated by output validator during the previous invocation of the submission.
> 
> To signal that the submission should be run again, the output validator must exit with code `42` and output the new input in the file `nextpass.in` in the feedback directory. Judging stops if no nextpass.in was created, or the output validator exited with any other code.
> 
> It is a judge error to create the `nextpass.in` file and exit with any other code than 42.

- `multipass` can be combined with `validation` set to `custom` or `custom interactive`.
- Regarding the timelimit, it should be checked per invocation of the submission.
- The `judgemessage.txt` and `judgeerror.txt` should not be removed between invocations of the output validator i.e. the validator is responsible for appending/overwriting to the files in there

Comments:
- If the validator uses `stdout` and `stderr` you likely want to display that per invocation

Open questions:
- Can the output validator (inside feedback?) write other files that persist between invocations?

Are there more things that need to be properly defined?

**Comment by mzuenni** [comment]
Maybe these are helpful for testing:
https://github.com/RagnarGrootKoerkamp/BAPCtools/tree/feat/multipass/test/problems/multipass
https://github.com/RagnarGrootKoerkamp/BAPCtools/tree/feat/multipass/test/problems/interactivemultipass

**Comment by meisterT** [comment]
Example interactive problem on Kattis: https://open.kattis.com/problems/guessinggame2

**Comment by meisterT** [comment]
To summarize my understanding of the spec:
- the submission is first called with the given input
- the validator can decide whether another pass is needed
- if so, it needs to exit with 42, and put the new input in `feedback/nextpass.in`
- if no new input is created, then the typical 23/42 exit codes indicate WA/AC

IMHO, the spec should be extended to include a  required upper bound of passes

The following changes are likely required in DOMjudge:
- in the judging backend, we probably could get away with some modifications to https://github.com/DOMjudge/domjudge/blob/main/judge/testcase_run.sh
- in the database/schema, we have to think about how to represent the multiple answers, probably having a list of https://github.com/DOMjudge/domjudge/blob/main/webapp/src/Entity/JudgingRun.php per judging
- in the UI, we have to see, how to represent things in the UI
- no actual scoring changes

All in all doable, but still touches more than just the backend.

**Comment by mzuenni** [comment]
I think an upper bound was not considered (it is intended that the number of passes is variable so the validator is responsible for deciding the number of passes)... still an upper bound might be a good safeguard against bugs in the validator that result in an infinite loop... but not sure if that should be part of the spec or just a domjudge thing (is the maximum runtime of the output validator specified for normal problem?).

**Comment by eldering** [comment]
A limit of validation passes was added in https://github.com/Kattis/problem-package-format/pull/231, defaulting to `2` if nothing is specified.

**Comment by meisterT** [comment]
We discussed this in person a little bit more.

For a complete implementation we want to do/consider the following:
- In the backend, make multipass problems the default, where non-multipass problems are just a multipass problem with one pass.
- That implies that we do the same isolation between passes that we do between test case runs in terms of chroot and cgroups that we do today.
- From an entity / data storage perspective, we likely want to have the same data about each pass, that we currently have in a `JudgingRun`. The easiest might just be to say that a judging run is the run on a single pass vs a single test case.
- We have to design how to display multipass output nicely in the UI. We have to consider that the output could be in binary format, so we should be able to handle this as well.
- For the API, the most likely thing is to add a `pass` type, which probably will look very very similar to https://ccs-specs.icpc.io/draft/contest_api#runs and is a child of it.
- This would also allow downstream clients to continue to work transparently as we don't modify existing structures and it is a backwards-compatible change (with additional info) on top.
- We probably should use this opportunity to better document the flow in the judging backend.

Implementation wise we probably can do baby steps and learn as we go and don't **need** to do any UI/API/storage changes in the first step as we can just display the info from the last pass.

**Comment by meisterT** [comment]
@mzuenni for the wrong answers in https://github.com/RagnarGrootKoerkamp/BAPCtools/tree/master/test/problems/multipass/submissions/wrong_answer it would be great if you would add comments on what to expect (first pass fail or second pass fail).

**Comment by mzuenni** [comment]
I added a comment for each submission

**Comment by cubercsl** [comment]
The latest changes for multi-pass problem support seem that we can only change the problem type by importing the problem package.

However, we cannot change the problem type (interactive or multi-pass) and `MultipassLimit` on UI.

**Comment by nickygerritsen** [comment]
Feel free to open a PR. Otherwise we will do it soon ™️

**Comment by nickygerritsen** [comment]
It should work but there is no UI for it yet. We would love some testing for sure.
The backend code is quite simple so I'm not very worried it wont work

**Comment by meisterT** [comment]
Agreed with Nicky. It works but we currently do not show any intermediate output or generated input in the web interface. Just the last pass is uploaded, either the actual last pass if successful or an intermediate one because the submission already failed. 

Testing beforehand is always appreciated though 😀

**Comment by mzuenni** [comment]
multipass output validators might try to read/overwrite from the `judgemessage.txt` so it should be the same file.

A typical usecase is to write the input of the first pass to some file in the feedback dir so that the second pass knows the input of the first pass and the input of the second pass.

**Comment by mzuenni** [comment]
yea the good practice here is to write this to any other file in the feedback dir (and thats also the reason why those files must persist between runs) but the spec says that all files in that dir should persist this allows to simply reuse the same directory across all passes.

**Comment by meisterT** [comment]
I assume the spec is not set in stone 🤪

**Comment by mzuenni** [comment]
Yes :D but linking all files that were created in the feedback dir (or even copying) sounds more complicating than just linking the dir as a whole?

**Comment by meisterT** [comment]
We currently also add validator stdout / stderr to the judgemessage file: https://github.com/DOMjudge/domjudge/blob/main/judge/testcase_run.sh#L270

See also the discussion we had here: https://github.com/RagnarGrootKoerkamp/BAPCtools/pull/393#discussion_r1779651529

Of course this could be changed as well, but it feels a lot less hacky to me to communicate between passes through a separate file.

**Comment by mpsijm** [comment]
I've been able to run a (still confidential) multi-pass problem that assumes the feedback dir is "the same" between every pass on my local DOMjudge dev environment using these changes: https://github.com/mpsijm/domjudge/commit/e2de765d200435c90cba78e0e0c166487839266b

Note that "the same" is here very hackily implemented by copying the feedback dir from the previous pass and removing `nextpass.in` (because if our output validator does not write to `nextpass.in`, DOMjudge thought that there would be more passes because of the existence of the previous `nextpass.in` file) :stuck_out_tongue:

**Comment by meisterT** [comment]
I would prefer that we standardize on using a single file, such as `validator.out`. This file would then be copied to `validator.in` for the next pass, with no other files being copied between passes.

I do think this would not limit what a validator could do / what problems you could create, but it's cleaner.

@mpsijm @mzuenni WDYT?

**Comment by mpsijm** [comment]
I don't have much experience with multi-pass problems yet, and I haven't been following the discussions on the problem spec, but if this would be the way to standardize, this suggestion should probably be discussed in https://github.com/Kattis/problem-package-format. Nonetheless, it does sound like a clean solution :slightly_smiling_face:

**Comment by mzuenni** [comment]
`validator.out` would be sufficient but I see at least 3 use cases where multiple files are preferable:

- to keep the information of different passes in different files
- to be able to directly byte dump/read data structures to different files (makes reading and writing a lot easier)
- to allow "libraries" to have their own files that they write to without interfering with what the validator does

**Comment by meisterT** [comment]
cc @nickygerritsen @eldering as you probably want to discuss the spec as part of a clics call as I don't think it has been discussed before or thought through by anyone of DOMjudge/PC2 (e.g. https://github.com/Kattis/problem-package-format/pull/168 doesn't seem to have seen any discussion)

My personal point of view is to keep it as simple (read: restricted) as possible to be able to reason about both from a system and a jury perspective. But it might also be that I lack imagination on multi-pass problems in general.

**Comment by nickygerritsen** [comment]
I believe Jaap was part of the discussion.
Clics is very silent currently but we will make sure it gets discussed when it starts again.

**Comment by mzuenni** [comment]
> > to keep the information of different passes in different files
> 
> So your validator needs info from not just the last pass, but the second to last one?

yes for example or last pass and first pass.

**Comment by eldering** [comment]
@kunyavskiy mentioned https://ucup.ac/scoreboard/39 as a contest with many/only multipass problems, which we might use some from to test our implementation. Attached the problem statements.
[statements-2-16.pdf](https://github.com/user-attachments/files/18156832/statements-2-16.pdf)

**Comment by meisterT** [comment]
@mzuenni what do you think about Ragnar's proposal to keep `${feedbackdir}/multipass` between runs but nothing else? (@mpsijm already gave his 👍 )

@GassaFM Are the problems for the above contest available as ICPC problem packages (some form of https://icpc.io/problem-package-format/spec/2023-07-draft.html)?

**Comment by mzuenni** [comment]
yes this is also fine for me (as long as its guaranteed that the dir already exists).

**Comment by GassaFM** [comment]
> [@GassaFM](https://github.com/GassaFM) Are the problems for the above contest available as ICPC problem packages (some form of https://icpc.io/problem-package-format/spec/2023-07-draft.html)?

First time I see this format, so no :) but convertible, most likely.

They are in the same format as Petrozavodsk Camp archives, or Andrew Stankevich contests, or older NERC contests, or some such. The format is used in NERC and around for 25+ years, now partly superseded by Polygon packages.

Kept in a private Subversion repository. I can copy that to a private Github and share with your team. (In my opinion, _public_ Github would be a bit too visible to archive solvers.)

**Comment by mzuenni** [comment]
you can also use this for testing purposes: https://github.com/RagnarGrootKoerkamp/BAPCtools/tree/multipassdir/test/problems/alternativeencryption

There are some submissions under `submission_disabled` that write files which domjudge should be able to handle (BAPCtools does not).

**Comment by cubercsl** [comment]
I saw that https://github.com/DOMjudge/domjudge/commit/cd8d38c59cc0c797a605aa5cc8f884b4979939ab
documented that `Multi-pass problems cannot be combined with interactive or submit-answer problem types`. 
But in the problem format https://www.kattis.com/problem-package-format/spec/2025-09.html#type the multi-pass is not incompatible with interactive and BAPCTools has an example https://github.com/RagnarGrootKoerkamp/BAPCtools/tree/main/test%2Fproblems%2Finteractivemultipass .

DOMjudge did not prevent this behavior either (at least not in the 9.0 release, it can still set by UI). However, the corresponding implementation is missing when importing from  zip. 
But judging seems good.

**Comment by meisterT** [comment]
@cubercsl I wanted to be conservative in documenting what we currently support. We will likely make it work and officially support in the future.
